File: app/src/main/java/com/rmuhamed/sample/poketest/config/AppConfigurator.java
package com.rmuhamed.sample.poketest.config;

import android.content.Context;
import org.jetbrains.annotations.NotNull;

public class AppConfigurator {
    private static boolean INITIALIZED;

    private final AppDatabase appDatabase;
    private final RestApiDefinition apiDefinition;

    private AppConfigurator(Context context, String baseUrl) {
        if (INITIALIZED) {
            throw new RuntimeException("Already initialized App Configuration");
        } else {
            INITIALIZED = true;

            this.appDatabase = DatabaseConfigurator.getInstance(context);
            this.apiDefinition = RestApiConfigurator.createApi(baseUrl);
        }
    }

    public static AppConfigurator initialize(@NotNull Context context, @NotNull String baseUrl) {
        return new AppConfigurator(context, baseUrl);
    }


    public RestApiDefinition getApiDefinition() {
        return apiDefinition;
    }

    public AppDatabase getAppDatabase() {
        return appDatabase;
    }
}

File: app/src/main/java/com/rmuhamed/sample/poketest/config/AppConfigurator.kt
package com.rmuhamed.sample.poketest.config

import android.content.Context
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

class AppConfigurator private constructor(context: Context, baseUrl: String) {

    val appDatabase: AppDatabase
    val apiDefinition: RestApiDefinition
    val threadExecutor: ExecutorService

    init {
        if (INITIALIZED) {
            throw RuntimeException("Already initialized App Configuration")
        } else {
            INITIALIZED = true

            this.appDatabase = DatabaseConfigurator.createDatabase(context)
            this.apiDefinition = RestApiConfigurator.createApi(baseUrl)
            this.threadExecutor = Executors.newFixedThreadPool(4)
        }
    }

    companion object {
        private var INITIALIZED: Boolean = false

        fun initialize(context: Context, baseUrl: String): AppConfigurator {
            return AppConfigurator(context, baseUrl)
        }
    }
}
